src/hotspot/share/gc/shared/barrierSetNMethod.cpp

```diff
@@ -172,6 +178,9 @@ int BarrierSetNMethod::nmethod_stub_entry_barrier(address* return_address_ptr) {
   nmethod* nm = cb->as_nmethod();
   BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
 
+  // Check for disarmed method here to avoid going into DeoptimizeNMethodBarriersALot code
+  // too often. nmethod_entry_barrier checks for disarmed status itself,
+  // but we have no visibility into whether the barrier acted or not.
   if (!bs_nm->is_armed(nm)) {
```

Comment:
I would still like this check gone, together with the comment above.

Reply:
I agree with Neethu, though: I think we should proceed to the `DeoptimizeNMethodBarrierALot` block and cross-modify fences only when barrier acted. We know from testing that it hurts otherwise. I would prefer this change not to introduce new performance potholes, even for fastdebug modes.

If the argument is cleanliness on who is checking "armed", and that we decide it should be solely in backend, then the middle ground might be adding the out-parameter, like `nmethod_entry_barrier(nmethod* nm, bool has_acted)`, and checking that before proceeding here? That feels uglier than just leaving the check here.